File: Den.Dev.Grunt/Den.Dev.Grunt/Core/Modules/ModuleBase.cs
// <copyright file="ModuleBase.cs" company="Den Delimarsky">
// Developed by Den Delimarsky.
// Den Delimarsky licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
// The underlying API powering Den.Dev.Grunt is managed by Halo Studios and Microsoft. This wrapper is not endorsed by Halo Studios or Microsoft.
// </copyright>

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Den.Dev.Grunt.Core.Foundation;
using Den.Dev.Grunt.Endpoints;
using Den.Dev.Grunt.Models;

namespace Den.Dev.Grunt.Core.Modules
{
    /// <summary>
    /// Base class for all API modules providing shared functionality for making HTTP requests.
    /// </summary>
    public abstract class ModuleBase
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="ModuleBase"/> class.
        /// </summary>
        /// <param name="client">The client instance to use for API requests.</param>
        /// <param name="origin">The origin/subdomain for this module's endpoints.</param>
        protected ModuleBase(ClientBase client, string origin)
        {
            this.Client = client ?? throw new ArgumentNullException(nameof(client));
            this.Origin = origin ?? throw new ArgumentNullException(nameof(origin));
        }

        /// <summary>
        /// Gets the client instance used for executing API requests.
        /// </summary>
        protected ClientBase Client { get; }

        /// <summary>
        /// Gets the origin/subdomain for this module's API endpoints.
        /// </summary>
        protected string Origin { get; }

        /// <summary>
        /// Validates that a value is not null.
        /// </summary>
        /// <typeparam name="T">The type of the value.</typeparam>
        /// <param name="value">The value to check.</param>
        /// <param name="paramName">The parameter name for the exception.</param>
        /// <exception cref="ArgumentNullException">Thrown when value is null.</exception>
        protected static void ValidateNotNull<T>(T value, string paramName)
            where T : class
        {
            if (value is null)
            {
                throw new ArgumentNullException(paramName);
            }
        }

        /// <summary>
        /// Validates that an integer is within a specified range.
        /// </summary>
        /// <param name="value">The value to check.</param>
        /// <param name="min">The minimum allowed value (inclusive).</param>
        /// <param name="max">The maximum allowed value (inclusive).</param>
        /// <param name="paramName">The parameter name for the exception.</param>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when value is outside the range.</exception>
        protected static void ValidateRange(int value, int min, int max, string paramName)
        {
            if (value < min || value > max)
            {
                throw new ArgumentOutOfRangeException(paramName, $"Value must be between {min} and {max}.");
            }
        }

        /// <summary>
        /// Builds a full URL from a relative path using this module's origin.
        /// </summary>
        /// <param name="path">The relative path (should start with /).</param>
        /// <returns>The fully qualified URL.</returns>
        protected string BuildUrl(string path) =>
            $"https://{this.Origin}.{HaloCoreEndpoints.ServiceDomain}{path}";

        /// <summary>
        /// Executes a GET request against the API.
        /// </summary>
        /// <typeparam name="T">The expected response type.</typeparam>
        /// <param name="path">The relative API path.</param>
        /// <param name="useClearance">Whether to include the clearance token.</param>
        /// <param name="useSpartanToken">Whether to include the Spartan token. Defaults to true.</param>
        /// <returns>The API response container.</returns>
        protected Task<HaloApiResultContainer<T, RawResponseContainer>> GetAsync<T>(
            string path,
            bool useClearance = false,
            bool useSpartanToken = true) =>
            this.Client.ExecuteAPIRequest<T>(
                this.BuildUrl(path),
                HttpMethod.Get,
                useSpartanToken,
                useClearance,
                includeRawResponse: this.Client.IncludeRawResponses);

        /// <summary>
        /// Executes a GET request against a fully specified URL.
        /// </summary>
        /// <typeparam name="T">The expected response type.</typeparam>
        /// <param name="fullUrl">The full URL to request.</param>
        /// <param name="useClearance">Whether to include the clearance token.</param>
        /// <param name="useSpartanToken">Whether to include the Spartan token. Defaults to true.</param>
        /// <param name="customHeaders">Optional custom headers to include.</param>
        /// <param name="enforceSuccess">Whether to enforce success response codes.</param>
        /// <returns>The API response container.</returns>
        protected Task<HaloApiResultContainer<T, RawResponseContainer>> GetAsyncFullUrl<T>(
            string fullUrl,
            bool useClearance = false,
            bool useSpartanToken = true,
            List<KeyValuePair<string, string>>? customHeaders = null,
            bool enforceSuccess = true) =>
            this.Client.ExecuteAPIRequest<T>(
                fullUrl,
                HttpMethod.Get,
                useSpartanToken,
                useClearance,
                includeRawResponse: this.Client.IncludeRawResponses,
                customHeaders: customHeaders,
                enforceSuccess: enforceSuccess);

        /// <summary>
        /// Executes a POST request against the API.
        /// </summary>
        /// <typeparam name="T">The expected response type.</typeparam>
        /// <param name="path">The relative API path.</param>
        /// <param name="content">The JSON content to send.</param>
        /// <param name="useClearance">Whether to include the clearance token.</param>
        /// <param name="useSpartanToken">Whether to include the Spartan token. Defaults to true.</param>
        /// <returns>The API response container.</returns>
        protected Task<HaloApiResultContainer<T, RawResponseContainer>> PostAsync<T>(
            string path,
            string content = "",
            bool useClearance = false,
            bool useSpartanToken = true) =>
            this.Client.ExecuteAPIRequest<T>(
                this.BuildUrl(path),
                HttpMethod.Post,
                useSpartanToken,
                useClearance,
                content,
                includeRawResponse: this.Client.IncludeRawResponses);

        /// <summary>
        /// Executes a POST request with JSON serialized body.
        /// </summary>
        /// <typeparam name="T">The expected response type.</typeparam>
        /// <typeparam name="TBody">The type of the request body.</typeparam>
        /// <param name="path">The relative API path.</param>
        /// <param name="body">The object to serialize as JSON.</param>
        /// <param name="useClearance">Whether to include the clearance token.</param>
        /// <param name="useSpartanToken">Whether to include the Spartan token. Defaults to true.</param>
        /// <param name="contentType">The content type for the request.</param>
        /// <returns>The API response container.</returns>
        protected Task<HaloApiResultContainer<T, RawResponseContainer>> PostJsonAsync<T, TBody>(
            string path,
            TBody body,
            bool useClearance = false,
            bool useSpartanToken = true,
            APIContentType contentType = APIContentType.Json) =>
            this.Client.ExecuteAPIRequest<T>(
                this.BuildUrl(path),
                HttpMethod.Post,
                useSpartanToken,
                useClearance,
                JsonSerializer.Serialize(body),
                contentType: contentType,
                includeRawResponse: this.Client.IncludeRawResponses);

        /// <summary>
        /// Executes a PUT request against the API.
        /// </summary>
        /// <typeparam name="T">The expected response type.</typeparam>
        /// <param name="path">The relative API path.</param>
        /// <param name="content">The JSON content to send.</param>
        /// <param name="useClearance">Whether to include the clearance token.</param>
        /// <param name="useSpartanToken">Whether to include the Spartan token. Defaults to true.</param>
        /// <returns>The API response container.</returns>
        protected Task<HaloApiResultContainer<T, RawResponseContainer>> PutAsync<T>(
            string path,
            string content = "",
            bool useClearance = false,
            bool useSpartanToken = true) =>
            this.Client.ExecuteAPIRequest<T>(
                this.BuildUrl(path),
                HttpMethod.Put,
                useSpartanToken,
                useClearance,
                content,
                includeRawResponse: this.Client.IncludeRawResponses);

        /// <summary>
        /// Executes a PUT request with JSON serialized body.
        /// </summary>
        /// <typeparam name="T">The expected response type.</typeparam>
        /// <typeparam name="TBody">The type of the request body.</typeparam>
        /// <param name="path">The relative API path.</param>
        /// <param name="body">The object to serialize as JSON.</param>
        /// <param name="useClearance">Whether to include the clearance token.</param>
        /// <param name="useSpartanToken">Whether to include the Spartan token. Defaults to true.</param>
        /// <returns>The API response container.</returns>
        protected Task<HaloApiResultContainer<T, RawResponseContainer>> PutJsonAsync<T, TBody>(
            string path,
            TBody body,
            bool useClearance = false,
            bool useSpartanToken = true) =>
            this.Client.ExecuteAPIRequest<T>(
                this.BuildUrl(path),
                HttpMethod.Put,
                useSpartanToken,
                useClearance,
                JsonSerializer.Serialize(body),
                includeRawResponse: this.Client.IncludeRawResponses);

        /// <summary>
        /// Executes a PATCH request against the API.
        /// </summary>
        /// <typeparam name="T">The expected response type.</typeparam>
        /// <param name="path">The relative API path.</param>
        /// <param name="content">The JSON content to send.</param>
        /// <param name="useClearance">Whether to include the clearance token.</param>
        /// <param name="useSpartanToken">Whether to include the Spartan token. Defaults to true.</param>
        /// <returns>The API response container.</returns>
        protected Task<HaloApiResultContainer<T, RawResponseContainer>> PatchAsync<T>(
            string path,
            string content = "",
            bool useClearance = false,
            bool useSpartanToken = true) =>
            this.Client.ExecuteAPIRequest<T>(
                this.BuildUrl(path),
                HttpMethod.Patch,
                useSpartanToken,
                useClearance,
                content,
                includeRawResponse: this.Client.IncludeRawResponses);

        /// <summary>
        /// Executes a PATCH request with JSON serialized body.
        /// </summary>
        /// <typeparam name="T">The expected response type.</typeparam>
        /// <typeparam name="TBody">The type of the request body.</typeparam>
        /// <param name="path">The relative API path.</param>
        /// <param name="body">The object to serialize as JSON.</param>
        /// <param name="useClearance">Whether to include the clearance token.</param>
        /// <param name="useSpartanToken">Whether to include the Spartan token. Defaults to true.</param>
        /// <returns>The API response container.</returns>
        protected Task<HaloApiResultContainer<T, RawResponseContainer>> PatchJsonAsync<T, TBody>(
            string path,
            TBody body,
            bool useClearance = false,
            bool useSpartanToken = true) =>
            this.Client.ExecuteAPIRequest<T>(
                this.BuildUrl(path),
                HttpMethod.Patch,
                useSpartanToken,
                useClearance,
                JsonSerializer.Serialize(body),
                includeRawResponse: this.Client.IncludeRawResponses);

        /// <summary>
        /// Executes a DELETE request against the API.
        /// </summary>
        /// <typeparam name="T">The expected response type.</typeparam>
        /// <param name="path">The relative API path.</param>
        /// <param name="useClearance">Whether to include the clearance token.</param>
        /// <param name="useSpartanToken">Whether to include the Spartan token. Defaults to true.</param>
        /// <returns>The API response container.</returns>
        protected Task<HaloApiResultContainer<T, RawResponseContainer>> DeleteAsync<T>(
            string path,
            bool useClearance = false,
            bool useSpartanToken = true) =>
            this.Client.ExecuteAPIRequest<T>(
                this.BuildUrl(path),
                HttpMethod.Delete,
                useSpartanToken,
                useClearance,
                includeRawResponse: this.Client.IncludeRawResponses);
    }
}

File: src/node/src/modules/base/module-base.ts
import type { ClientBase } from '../../clients/base/client-base';
import type { HaloApiResult } from '../../models/common/api-result';
import { ApiContentType } from '../../models/common/api-content-type';
import { buildServiceUrl } from '../../endpoints/halo-core-endpoints';

/**
 * Abstract base class for all Halo Infinite API modules.
 *
 * Provides shared functionality for building URLs and making HTTP requests.
 * Each module (Stats, Economy, GameCms, etc.) extends this class and uses
 * the helper methods to interact with its specific API endpoints.
 *
 * @example
 * ```typescript
 * class StatsModule extends ModuleBase {
 *   constructor(client: ClientBase) {
 *     super(client, HALO_CORE_ENDPOINTS.STATS_ORIGIN);
 *   }
 *
 *   async getMatchStats(matchId: string) {
 *     return this.get<MatchStats>(`/hi/matches/${matchId}/stats`);
 *   }
 * }
 * ```
 */
export abstract class ModuleBase {
  /**
   * Reference to the parent client for making HTTP requests.
   */
  protected readonly client: ClientBase;

  /**
   * Service origin for this module (e.g., 'halostats', 'economy').
   */
  protected readonly origin: string;

  /**
   * Creates a new module instance.
   *
   * @param client - Parent client instance
   * @param origin - Service origin for URL building
   */
  constructor(client: ClientBase, origin: string) {
    this.client = client;
    this.origin = origin;
  }

  /**
   * Build a full URL from a relative path using this module's origin.
   *
   * @param path - API path starting with /
   * @returns Full HTTPS URL
   */
  protected buildUrl(path: string): string {
    return buildServiceUrl(this.origin, path);
  }

  /**
   * Execute a GET request to this module's service.
   *
   * @template T - Expected response type
   * @param path - API path (e.g., '/hi/players/xuid(...)/matches')
   * @param options - Request options
   * @returns Promise with the API result
   */
  protected get<T>(
    path: string,
    options: {
      useClearance?: boolean;
      useSpartanToken?: boolean;
      customHeaders?: Record<string, string>;
    } = {}
  ): Promise<HaloApiResult<T>> {
    return this.client.executeRequest<T>(this.buildUrl(path), 'GET', {
      useSpartanToken: options.useSpartanToken ?? true,
      useClearance: options.useClearance ?? false,
      customHeaders: options.customHeaders,
    });
  }

  /**
   * Execute a GET request to an absolute URL.
   *
   * Used when the URL doesn't follow the standard origin pattern
   * (e.g., blob storage URLs).
   *
   * @template T - Expected response type
   * @param fullUrl - Complete URL to request
   * @param options - Request options
   * @returns Promise with the API result
   */
  protected getFullUrl<T>(
    fullUrl: string,
    options: {
      useClearance?: boolean;
      useSpartanToken?: boolean;
      customHeaders?: Record<string, string>;
      enforceSuccess?: boolean;
    } = {}
  ): Promise<HaloApiResult<T>> {
    return this.client.executeRequest<T>(fullUrl, 'GET', {
      useSpartanToken: options.useSpartanToken ?? true,
      useClearance: options.useClearance ?? false,
      customHeaders: options.customHeaders,
      enforceSuccess: options.enforceSuccess ?? true,
    });
  }

  /**
   * Execute a POST request to this module's service.
   *
   * @template T - Expected response type
   * @param path - API path
   * @param body - Optional request body as string
   * @param options - Request options
   * @returns Promise with the API result
   */
  protected post<T>(
    path: string,
    body?: string,
    options: {
      useClearance?: boolean;
      useSpartanToken?: boolean;
      customHeaders?: Record<string, string>;
    } = {}
  ): Promise<HaloApiResult<T>> {
    return this.client.executeRequest<T>(this.buildUrl(path), 'POST', {
      useSpartanToken: options.useSpartanToken ?? true,
      useClearance: options.useClearance ?? false,
      body,
      customHeaders: options.customHeaders,
    });
  }

  /**
   * Execute a POST request with a JSON body.
   *
   * @template T - Expected response type
   * @template TBody - Request body type
   * @param path - API path
   * @param body - Request body (will be serialized to JSON)
   * @param options - Request options
   * @returns Promise with the API result
   */
  protected postJson<T, TBody>(
    path: string,
    body: TBody,
    options: {
      useClearance?: boolean;
      useSpartanToken?: boolean;
      contentType?: ApiContentType;
      customHeaders?: Record<string, string>;
    } = {}
  ): Promise<HaloApiResult<T>> {
    return this.client.executeRequest<T>(this.buildUrl(path), 'POST', {
      useSpartanToken: options.useSpartanToken ?? true,
      useClearance: options.useClearance ?? false,
      body: JSON.stringify(body),
      contentType: options.contentType ?? ApiContentType.Json,
      customHeaders: options.customHeaders,
    });
  }

  /**
   * Execute a PUT request with a JSON body.
   *
   * @template T - Expected response type
   * @template TBody - Request body type
   * @param path - API path
   * @param body - Request body (will be serialized to JSON)
   * @param options - Request options
   * @returns Promise with the API result
   */
  protected putJson<T, TBody>(
    path: string,
    body: TBody,
    options: {
      useClearance?: boolean;
      useSpartanToken?: boolean;
      customHeaders?: Record<string, string>;
    } = {}
  ): Promise<HaloApiResult<T>> {
    return this.client.executeRequest<T>(this.buildUrl(path), 'PUT', {
      useSpartanToken: options.useSpartanToken ?? true,
      useClearance: options.useClearance ?? false,
      body: JSON.stringify(body),
      customHeaders: options.customHeaders,
    });
  }

  /**
   * Execute a PATCH request with a JSON body.
   *
   * @template T - Expected response type
   * @template TBody - Request body type
   * @param path - API path
   * @param body - Request body (will be serialized to JSON)
   * @param options - Request options
   * @returns Promise with the API result
   */
  protected patchJson<T, TBody>(
    path: string,
    body: TBody,
    options: {
      useClearance?: boolean;
      useSpartanToken?: boolean;
      customHeaders?: Record<string, string>;
    } = {}
  ): Promise<HaloApiResult<T>> {
    return this.client.executeRequest<T>(this.buildUrl(path), 'PATCH', {
      useSpartanToken: options.useSpartanToken ?? true,
      useClearance: options.useClearance ?? false,
      body: JSON.stringify(body),
      customHeaders: options.customHeaders,
    });
  }

  /**
   * Execute a DELETE request.
   *
   * @template T - Expected response type
   * @param path - API path
   * @param options - Request options
   * @returns Promise with the API result
   */
  protected delete<T>(
    path: string,
    options: {
      useClearance?: boolean;
      useSpartanToken?: boolean;
      customHeaders?: Record<string, string>;
    } = {}
  ): Promise<HaloApiResult<T>> {
    return this.client.executeRequest<T>(this.buildUrl(path), 'DELETE', {
      useSpartanToken: options.useSpartanToken ?? true,
      useClearance: options.useClearance ?? false,
      customHeaders: options.customHeaders,
    });
  }

  /**
   * Validate that a parameter is not null or undefined.
   *
   * @param value - Value to check
   * @param paramName - Parameter name for error message
   * @throws Error if value is null or undefined
   */
  protected assertNotNull<T>(
    value: T | null | undefined,
    paramName: string
  ): asserts value is T {
    if (value == null) {
      throw new Error(`${paramName} cannot be null or undefined`);
    }
  }

  /**
   * Validate that a number is within a range.
   *
   * @param value - Value to check
   * @param min - Minimum allowed value (inclusive)
   * @param max - Maximum allowed value (inclusive)
   * @param paramName - Parameter name for error message
   * @throws RangeError if value is out of range
   */
  protected assertRange(
    value: number,
    min: number,
    max: number,
    paramName: string
  ): void {
    if (value < min || value > max) {
      throw new RangeError(`${paramName} must be between ${min} and ${max}`);
    }
  }

  /**
   * Validate that a string is not empty.
   *
   * @param value - Value to check
   * @param paramName - Parameter name for error message
   * @throws Error if value is empty
   */
  protected assertNotEmpty(value: string, paramName: string): void {
    if (!value || value.trim().length === 0) {
      throw new Error(`${paramName} cannot be empty`);
    }
  }
}
